crates/telio-starcast/src/multicaster.rs

```diff
@@ -0,0 +1,252 @@
+use futures::future::join_all;
+use pnet_packet::ipv4::{checksum, MutableIpv4Packet};
+use pnet_packet::udp::MutableUdpPacket;
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+use std::net::{IpAddr, Ipv4Addr, Ipv6Addr};
+use std::{net::SocketAddr, sync::Arc};
+use telio_sockets::SocketPool;
+use telio_task::io::chan::Rx;
+use telio_utils::{telio_log_debug, telio_log_error, telio_log_warn};
+use telio_wg::{DynamicWg, WireGuard};
+use tokio::net::UdpSocket;
+use tokio::task::JoinHandle;
+
+#[derive(Clone)]
+pub struct MulticasterIp {
+    pub ipv4: Ipv4Addr,
+    pub ipv6: Ipv6Addr,
+}
+
+#[cfg(feature = "test-util")]
+const MULTICAST_TRANSPORT_PORT: u16 = 0;
+#[cfg(not(feature = "test-util"))]
+const MULTICAST_TRANSPORT_PORT: u16 = 13569;
+const MAX_PACKET: usize = 2048;
+const IPV4_HEADER: usize = 20; // bytes
+
+/// Multicaster task that resends locally generated
+/// multicast traffic to other peers
+pub struct Multicaster {
+    wg: Arc<DynamicWg>,
+    multicast_socket: Arc<UdpSocket>,
+    local_socket: Arc<UdpSocket>,
+    packet_rx: Option<Rx<Vec<u8>>>,
+    forward_to_vpeer_task: Option<JoinHandle<()>>,
+    fanout_task: Option<JoinHandle<()>>,
+    multicaster_ip: Arc<MulticasterIp>,
+}
+
+impl Multicaster {
+    pub async fn new(
+        local_socket: Arc<UdpSocket>,
+        wg: Arc<DynamicWg>,
+        socket_pool: Arc<SocketPool>,
+        packet_rx: Rx<Vec<u8>>,
```

Comment:
so instead of receiving packet_rx and local socket, you should just get a singular `Chan<Vec<u8>>` that will be used for sending and receiving messages to and from @ThrasherLT MeshnetVirtualPeer